src/librustc_typeck/check/coercion.rs

```diff
@@ -692,12 +692,22 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
         debug!("coerce_from_fn_item(a={:?}, b={:?})", a, b);
 
         match b.kind {
-            ty::FnPtr(_) => {
+            ty::FnPtr(b_sig) => {
                 let a_sig = a.fn_sig(self.tcx);
                 // Intrinsics are not coercible to function pointers
                 if a_sig.abi() == Abi::RustIntrinsic || a_sig.abi() == Abi::PlatformIntrinsic {
                     return Err(TypeError::IntrinsicCast);
                 }
+
+                // Safe #[target_feature] functions are not assignable to safe fn pointers (RFC 2396)
+                if let ty::FnDef(def_id, _) = a.kind {
+                    if b_sig.unsafety() == hir::Unsafety::Normal
+                        && !self.tcx.codegen_fn_attrs(def_id).target_features.is_empty()
+                    {
+                        return Err(TypeError::TargetFeatureCast);
```

Comment:
Diagnostics nit: saving the spans to the `target_feature` attributes would make for good diagnostics here.

Reply:
How do I get the spans? Do I need to modify the `codegen_fn_attrs` query?

Reply:
Modifying the query might cause perf regressions in the happy path (but maybe worth testing). Otherwise, you have the `def_id`, so you can get the attributes via `tcx.get_attrs(def_id)` and then filter those attributes for `sym::target_feature` and extract the spans of the attributes. It should be too complicated, but it would be good to extract this logic, including `if let ty::FnDef(...)`, to a function `check_target_feature_to_fnptr`.